File: revision/majority_element.py
from typing import List
def majorityElement(nums: List[int]) -> int:
    #Brute Force gives TLE
    #count = 1
    #for i in range(len(nums)):
    #    for j in range(1,len(nums)):
    #        if nums[i] == nums[j]:
    #            count+=1
    #    if count > len(nums)// 2:
    #        return nums[i]
    #    count = 0
    #Using Hash Map
    #myhashmap = dict()
    #for x in nums:
    #    myhashmap[x] = myhashmap.get(x, 0) + 1
    #for x in nums:l
    #    if myhashmap[x] > len(nums) // 2:
    #        return x
    #Moores Voting Algo
    candidate = count = 0
    for x in nums:
        if count == 0:
            candidate = x
        if x == candidate:
            count+=1
        else:
            count-=1
    return candidate
nums = list(map(int, input().split()))
print(majorityElement(nums)) 
        

File: revision/majority_element.cpp
#include<iostream>
#include<vector>
using namespace std;
void input(vector<int>&nums){
   for(auto &i : nums){
       cin >> i;
   }
}
void display(vector<int>nums){
    for(auto const i : nums){
        cout << i << ' ' ;
    }
    cout << endl;
}
int majorityElement(vector<int>&nums){
    //Moores Voting Algo
    int candidate = 0, count = 0;
    for(auto i : nums){
        if (count == 0){
            candidate = i;
        }
        if (i == candidate){
            count++;
        }
        else{
            count--;
        }
    }
    return candidate;
}
int main()
{
    int n;
    cin >> n;
    vector<int>nums(n);
    input(nums);
    cout << majorityElement(nums) << endl;
    return 0;
}